Incorporating LyA 1B changes into AMTL

LNapolitano opened the pull request:
This PR will contain the changes necessary to incorporate the LyA 1B numobs changes into the alt MTLs. Planning to add a new actiontype to the tiletracker file, that when encountered will run `desitarget.mtl.update_lya_1b` on the alt ledger directory.

LNapolitano committed:
Starting developement on incorporating LyA 1B changes into AMTL

LNapolitano committed:
Adding Handling for Lya1B Actiontype in loop_alt_ledger

LNapolitano committed:
Adding tiletracker updating for lya1b action

LNapolitano committed:
Ensuring LyA1B action is only run for obscon = dark ledgers.

LNapolitano committed:
Adding optional lya1b flag for lya qso numobs increase action. Fixing timestamp specification.

LNapolitano committed:
Adding handling for extension in desitarget.mtl.update_ledger calls

LNapolitano committed:
Adding Handling for maketwostyle kwarg in desitarget.io.read_targets_in_tiles() calls

LNapolitano committed:
Adding additional update_ledget call to update dark ledgers on a dark1b tile.

LNapolitano committed:
Adding reading and merging of dark and dark1b targets for dark1b tiles.

LNapolitano committed:
Modifying use of read_targets_in_tiles to pass two directories for dark1b/bright1b

LNapolitano committed:
Reverting a small change that broke file path setup for reproducing runs

LNapolitano committed:
Fixing dark1b/bright1b file specification for update actions.

LNapolitano committed:
Changes to update_alt_ledger for 1b tiles observed before lya1b changes on 20250721

LNapolitano committed:
syncing tiletracker update behavior with main branch

LNapolitano committed:
Temporarily removing desitarget version checks, no longer necessary

LNapolitano committed:
Updating file permissions to match main branch

LNapolitano committed:
Managing whitespace merge issue

LNapolitano committed:
Managing whitespace merge issue

LNapolitano committed:
Removing desitarget_version specification and associated code

LNapolitano committed:
Adjusting update_ledger logic for 1B programs

LNapolitano committed:
Updated to reflect make/update tiletracker changes.

LNapolitano committed:
Adding SingleNodeInitialize/SingleNoderun scripts from main

LNapolitano committed:
Modifying HPList to load from yaml file, rather than csv

LNapolitano committed:
Initial Add of yaml file actiontype, including as tiletracker action

LNapolitano committed:
Added initializeYAML_HPTracker function, which creates a per-obscon YAML file based on the file creation times of the real mtl ledgers

LNapolitano committed:
Adding new yaml file generation step to the SingleNode Initialize script

LNapolitano committed:
Initial add of YAML file check/regeneration in updateTileTracker

LNapolitano committed:
Adding some additional log statements, and moving the YAML file creation such that it's per realization.

LNapolitano committed:
Initial commit of new "add_new_ledgers" function which handles "addnew" actions to create new ledgers in an already existing set of alternate MTL ledgers.

LNapolitano committed:
Fixing type casting in initializeAlternateMTLs call and adding some new log print out

LNapolitano committed:
Adjusting yaml ledger tracker behavior for non-1b surveys. These want all ledgers under the 'initial' key.

LNapolitano committed:
Adding veto actions to makeTileTracker

LNapolitano committed:
Adding process_vetoes_altmtl function for veto reproduction

LNapolitano committed:
Minor fixes to process_vetoes_altmtl

LNapolitano committed:
Fixing bitflip type error, adding 2 second offset to veto actions to fix ordering and adding a minor helper function to fix zcat endianness issues with new desi environment.

LNapolitano committed:
Merge branch 'main' into LGN-LyA-1B

LNapolitano committed:
Cleaning up unused code

LNapolitano committed:
Changing YAML HP Tracker rewrite into an atomic-write

LNapolitano committed:
Moving endian conversion fn outside of update_alt_ledger. Adding early-exit if array is already big endian.

LNapolitano committed:
Re-adding zfix argument to update_alt_ledger

LNapolitano committed:
Merge branch 'main' into LGN-LyA-1B

LNapolitano committed:
Update fatools.py to match main branch FAVER logic

Removing special handling for tiles with 4.0 < FAVER < 5.0. Testing shows this is no longer necessary.

LNapolitano committed:
Adding a try/except wrapper to importing desitarget.mtl.update_lya_1b

This should altmtltools to be used in pre desitarget version 3.4.0 environments, as is currently desirable for the mock alt-mtls due to memory issues in later versions.

LNapolitano committed:
Merge branch 'main' into LGN-LyA-1B

LNapolitano committed:
Removing targets warning when running mocks. This is obselete functionality.

LNapolitano committed:
Removing resolved TODO comment

LNapolitano committed:
Changing default YAML-TileTracker read location (makeTileTracker and add_new_ledgers affected).

aureliocarnero committed:
Update DA2ALTMTLRealizationsBRIGHT_mock.sh

comment out target option since is not used for mocks

aureliocarnero committed:
Update DA2ALTMTLRealizationsDARK_holi.sh

Remove target option, is not used

aureliocarnero committed:
Update DA2ALTMTLRealizationsDARK_mock.sh

remove targfile. Is not needed in mocks

aureliocarnero committed:
Update Y1ALTMTLRealizationsBRIGHT_mock.sh

remove targfile for mocks. Is not used

aureliocarnero committed:
Update Y1ALTMTLRealizationsDARK_mock.sh

remove targfile option, is not needed

aureliocarnero committed:
Update Y1ALTMTLRealizationsDARK_mock_init.sh

remove targfile option, is not needed for mocks

aureliocarnero commented on bin/SingleNode-InitializeAltMTLsParallel.py:
Maybe initializeYAML_HPTracker is not needed any more here?

LNapolitano replied:
Agreed, thanks!

LNapolitano replied:
I've re-written this section by adding a new --yamlpath argument to the script, and forming the path to the yaml file using that and the obscon. 

https://github.com/desihub/LSS/pull/162/commits/140cc2f59d3b6cd0ff1e7abf1b76345bb81f128c
https://github.com/desihub/LSS/pull/162/commits/28924ce6b8f7b0c947ec92715abc9f1101b71afc

jalasker commented on bin/DA2ALTMTLRealizationsBRIGHT_mock.sh:
```diff
@@ -222,13 +222,9 @@ secondary=''
-#targfile='' #CHANGEME IF RUNNING ON MOCKS
-#targfile='--targfile=/global/cfs/cdirs/desi/target/catalogs/dr9/1.1.1/targets/main/resolve/' #Main survey target directory
+#ACR: we deactivate this option since is not used for mocks, it is not needed
+targfile='' 
```
-  Let’s remove the targfile variable from the code entirely at some point since only the mocks should have required this. Then we can remove the option and the commented associated options from the script files.
- Not going to make this comment for every bin/*.sh file but it applies to all of them

LNapolitano replied:
Agreed. I'll pass this along to @aureliocarnero as well for the mock scripts. Not urgent, but good to remove the targfile variables at some point.

jalasker commented on bin/SingleNode-InitializeAltMTLsParallel.py:
```diff
@@ -90,13 +91,9 @@
-    log.debug('First healpixel: {0:d}'.format(HPList[0]))
-    log.debug('Last healpixel: {0:d}'.format(HPList[-1]))
-    log.debug('Number of healpixels: {0:d}'.format(int(len(HPList))))
+# 20260401 LGN - Modifying HPList to be read from a YAML file, not a csv
```
Can we safely remove the legacy code and those comments now? It sounds like YAML is working and we always have prior tagged versions on GitHub in case we need to access it again.

LNapolitano replied:
Thanks James! Agreed, we can remove some of the comments here. I think the HPList info is still useful, as we do load that from the yaml file. Also I've just rewritten this section to use a --yamlpath argument to specify the directory name, with the default value supplied.

https://github.com/desihub/LSS/pull/162/commits/28924ce6b8f7b0c947ec92715abc9f1101b71afc

jalasker commented on bin/SingleNode-InitializeAltMTLsParallel.py:
```diff
@@ -138,8 +156,17 @@ def procFunc(nproc):
             log.info(args.finalDir)
-        log.info('shuffleSubpriorities: {0}'.format(args.shuffleSubpriorities))
+
+        # 20260420 LGN - Adding some additional log statements here for reproducibility
```
put these inside of an “if verbose and/or debug” block. It’s already hard to read the log files.

LNapolitano replied:
Done! https://github.com/desihub/LSS/pull/162/commits/dd5349aedf4c8a283e4ebfd8deb82486582c7db4

I just had to work out the tiletracker meta change first, as I was relying on these statements otherwise.

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -22,21 +22,36 @@
 
+#LGN 20260617 - Adding a T/E wrapping around the update_lya_1b import
+#             - This should enable the this code to be used with older desitarget versions.
+try:
```
It should be more reliable to directly check the version of desitarget rather than excepting all ImportErrors

LNapolitano replied:
Agreed, the try/except was a little nasty. I've replaced this with a "hasattr" check:
https://github.com/desihub/LSS/pull/162/commits/95576c2c9484b2d29fc7f4b622b0d1843478024b

This whole block is hopefully temporary as the mock team investigates the memory issues they encounter in desitarget versions after 3.0.0

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -163,7 +175,7 @@ def evaluateMask(bits, mask, evalMultipleBits = False):
 def flipBit(cat, bit2Flip, cond = None, fieldName = 'DESI_TARGET', mode = 'on'):
     #only works on single bits
-    assert( np.abs( np.log2(int(bit2Flip)) - int(np.log2(int(bit2Flip))) ) < 0.001  )
+    assert( np.abs( np.log2(int(bit2Flip)) - np.log2(int(bit2Flip)) ) < 0.001  )
```
IIRC, the purpose of this check was to ensure that the argument bit2Flip was a power of two. If you want to remove the “int()” from around the second argument of that assert statement, you should replace the entire argument with “bit2Flip == bit2Flip” since you’re only catching nans with that anyway.

LNapolitano replied:
Yeah, this is something that I believe broke with a new numpy version at some point? The log2 function required the explicit casting I believe. Anyway, since this is working for now I'm going to leave it as is, but I agree I can take another look later.

jalasker replied:
If the check here was solely for bit2Flip being a power of 2, then this answer is a lot smoother than what I wrote trying to avoid floating point rounding errors and vagaries of built-ins: https://stackoverflow.com/a/57025941 ((n & (n-1) == 0) and n != 0)

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -500,8 +512,40 @@ def updateTileTracker(altmtldir, endDate, survey = 'main', obscon = 'DARK'):
         print('New end date is the same as the current end date, update will not be performed')
         return
 
+    #only pass the lya1b keyword if prev_endDate is before 20250721 and new enddate is after 20250721
```
Anything related to parsing/creating the YAML files I’m going to assume is correct based on your testing since I’m not super familiar with the syntax. I want to get this review done quickly so you can start running the new AMTLs again. Let me know if you want me to review the YAML changes more closely.

LNapolitano replied:
Thanks! It should all work based on my testing.

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -676,6 +744,56 @@ def makeTileTracker(altmtldir, survey = 'main', obscon = 'DARK', startDate = Non
+
+    #LGN 07/29/25: adding new special action for the LyA QSO NUMOBS increase
+    #LGN This runs for all dark time surveys with actions at times later than 2025-07-21
+    #LGN unless the lya1b flag is set to false. Only change this flag if you are certain
```
This warning should be propagated  to the bin/*.sh files and/or written to log file/console if set to False.

LNapolitano replied:
Agreed, I've added some additional log statements if this doesn't run:
https://github.com/desihub/LSS/pull/162/commits/5344865178e660e06cd2e1b59c30b8abf7c4c687

This should be something we absolutely always want to be on. But it seemed best to have the option to disable it. Agreed we should propagate this down to the .sh scripts. I may need to do some extra wiring for that to be possible.

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -676,6 +744,56 @@ def makeTileTracker(altmtldir, survey = 'main', obscon = 'DARK', startDate = Non
+    if len(dates) > 1:
+        for date in dates[1:]:
+            #LGN 20260401: If date is in the timeframe between startDate and endDate add a new action for it
+            if startDate_str < date and endDate_str > date:
```
Shouldn’t one of these date checks be “<=“ or “>=“?

LNapolitano replied:
Definitely, thanks! I've made the endDate check into a >=

https://github.com/desihub/LSS/pull/162/commits/eb5aec789a5d1fb57fc81d7adaf01eb14cd61eae

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -706,7 +824,68 @@ def trimToMTL(notMTL, MTL, debug = False, verbose = False):
+    # LGN 20260407 First use xml to parse the svn log for the directory
+    # LGN 20260407 creating a dictionary mapping dates 'YYYY-MM-DD' to the added healpixels
+    directory = os.path.join(real_mtl_dir,obscon)
+    result = subprocess.run(
```
Flagging that we decided yesterday that we wanted to rewrite this without using svn directly in the script. A file written to a (DESI-)universally accessible directory in CFS sounded like a good option.

LNapolitano replied:
Yep! This is the function that writes those globally accessible YAML files. The change was to make it so that we don't need to run this function during initialization or when updating a set of ledgers. But we still need the function to generate the files in the first place.

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -744,6 +923,7 @@ def initializeAlternateMTLs(initMTL, outputMTL, nAlt = 2, genSubset = None, seed
     fn = initMTL.split('/')[-1]
     log.info('reading initial MTL(s)')
     allentries = Table.read(initMTL) 
+    log.info('initial MTL size: {}'.format(len(allentries)))
```
This, and many other log statements I left in this function, should be behind “if verbose” and/or “if debug” to clean up log files.

LNapolitano replied:
Added: https://github.com/desihub/LSS/pull/162/commits/669b5e630a6e6a1413d10e8acbbd2c2054a80d91

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -851,6 +1031,8 @@ def initializeAlternateMTLs(initMTL, outputMTL, nAlt = 2, genSubset = None, seed
             #makeTileTracker(outputMTLDir, survey = survey, obscon = obscon,overwrite = False,
             #startDate = startDateShort, endDate = endDateShort)
+        else:
+            log.info('tiletracker already exists, not overwriting')
```
“if verbose” and/or “if debug”

LNapolitano replied:
Added: https://github.com/desihub/LSS/pull/162/commits/669b5e630a6e6a1413d10e8acbbd2c2054a80d91

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -1051,7 +1233,9 @@ def initializeAlternateMTLs(initMTL, outputMTL, nAlt = 2, genSubset = None, seed
             #JL - reset TARGET_STATES based on new target bits. This step isn't necessary for AMTL function but makes debugging using target states vastly easier. 
             initialentries['TARGET_STATE'][ELGNewHIP & np.invert(QSOs)] = np.broadcast_to(np.array(['ELG_HIP|UNOBS']), np.sum(ELGNewHIP & np.invert(QSOs)  ) )
 
+        log.info('Initial Entries Size: {}'.format(len(initialentries)))
```
“if verbose” and/or “if debug”

LNapolitano replied:
Added: https://github.com/desihub/LSS/pull/162/commits/669b5e630a6e6a1413d10e8acbbd2c2054a80d91

jalasker commented on py/LSS/SV3/altmtltools.py:
```diff
@@ -1051,7 +1233,9 @@ def initializeAlternateMTLs(initMTL, outputMTL, nAlt = 2, genSubset = None, seed
 
+        log.info('Initial Entries Size: {}'.format(len(initialentries)))
         retval = desitarget.io.write_mtl(outputMTLDir, initialentries, survey=survey, obscon=obscon, extra=meta, nsidefile=meta['FILENSID'], hpxlist = [meta['FILEHPX']])
+        log.info('write_mtl return value: {}'.format(retval))
```
“if verbose” and/or “if debug”

LNapolitano replied:
Added: https://github.com/desihub/LSS/pull/162/commits/669b5e630a6e6a1413d10e8acbbd2c2054a80d91

jalasker commented on py/LSS/SV3/fatools.py:
```diff
@@ -383,18 +384,19 @@ def get_fba_fromnewmtl(tileid,mtldir=None,getosubp=False,outdir=None,faver=None,
-                    log.info(f'reproducing = {reproducing}') 
-            
+            #LGN 01/28/26: Adding compatibility with BRIGHT1B
+            elif targver == '1.0.0' or targver == '3.0.0' or targver == '3.2.0':
```
I’m a little worried about this line being applied to higher target versions than the default (1.1.1). These may(/will?) have new targets that aren’t present in 1.1.1 and so removing the check we have to see if v3.0.0 and 3.2.0 are being incorporated correctly feels suboptimal.

LNapolitano replied:
Yeah, I was admittedly a little unsure about this one. The intention here was to handle the very early survey tiles by directly copying their  targ files rather than generating them via `altcreate_mtl()` is that correct? 

Since this only ever runs with reproduction = True, it won't affect our standard reproduction = False runs. And I have verified that we can run dark1b/bright1b with reproduction = False. Maybe the best solution is to just remove 3.0.0 and 3.2.0 from this check and move them into the earlier statement:
`if (not reproducing) or (targver == '1.1.1'):`

But I'm not sure I totally understand the intention here either. Let me know what you think would be best.

jalasker replied:
The point I wanted to make here is related to the original reason this copy block was made for the reproduction runs. 

Originally there was a very short period of time (few days-weeks?) where DESI used target version 1.0.0 rather than 1.1.1. It proved to be challenging to exactly handle this target version change with the precision required for an exact reproduction test, but it was simple (IIRC we didn't actually have to make any changes) to get it working good enough from a statistical POV given the small number of tiles observed in that period of time. 

Therefore, we decided we could directly copy a handful of tiles and test that everything downstream of those tiles ran as expected and that would validate the remainder of the survey running on the same target version. 

With the version change going forward into BRIGHT1B, we have a similar scenario where 1.1.1 was used for a period of time and going forward we are using 3.0.0/3.2.0. Except this time we are doing the exact target file copy for the new version which will be used going forward. This means that any reproduction test should automatically succeed starting at the date where you use 3.0.0/3.2.0. At best, it means that no targets in v3.0.0 /3.2.0 are being properly tested, but the v1.1.1 targets which are still remaining get properly tested.

jalasker reviewed (commented):
Accidentally submitted the review as a comment, resubmitted all of the content from the comment as review comments directly on the relevant lines of code except for this comment in reply to one of your comments: 

> The addnew function is somewhat unique in that it requires some variables set at the creation of a set of alt mtl ledgers, namely {"seed", "reproducing", "shuffleSubpriorities", "shuffleBrightPriorities", "shuffleELGPriorities", "PromoteFracBGSFaint", "PromoteFracELG"}. Presently the code looks to find these in an initialization log file (default format: "Initialize{Obscon}AltMTLsParallelOutput_mainRepro.out"). We will need to discuss how to best handle accessing these keywords for the mock altMTLs.

- This is one of the reasons I liked having the script copied over from the initialization step. We don’t currently have any easily readable memory for this beyond the initial file (which, even with the copy checking step, can still be manually changed if it’s changed in the AMTL directory, so not perfect!). However, I believe we can store all of this information as metadata in the tile tracker ECSV header.

LNapolitano committed:
Removing initializeYAML_HPTracker() as files are now centralized, and do not need to be built for each realization.

LNapolitano committed:
Adding new --yamlpath argument, to locate the per-obscon yaml files. Removing legacy comments.

LNapolitano committed:
Adding initialization keywords to tiletracker meta info.

LNapolitano committed:
Adjusting add_new_ledgers to use the new tiletacker meta keywords.

LNapolitano committed:
Moving some log statements into an optional block (debug/verbose)

LNapolitano committed:
Replacing try/except wrapping for update_lya_1b in earlier desitarget versions with a "hasattr" check instead.

LNapolitano committed:
Adding informative log statements if lya1b action is not added to the tiletracker for DARK surveys.

LNapolitano committed:
Adding >= to add_new_ledger tiletracker creation check

LNapolitano committed:
Adding if (verbose or debug) checks to multiple log statements.

LNapolitano committed:
Changing get_fba_fromnewmtl to always call altcreate_mtl unless (reproducing = True and TARGVER = 1.0.0)

This ensures target files are generated using altcreate_mtl for DARK1B and BRIGHT1B fa actions

LNapolitano committed:
Adding block to read the targver and target directory directory from the fiberassign header info for main survey tiles.
